### PR TITLE
Add RetrieveChat

### LINKED ISSUE #1107: Auto FLAML code/api generation
Use LLM or LLM-based agents to generate FLAML code for users who need to perform automl or tuning tasks.  

May need more input from @thinkall
```[tasklist]
### Tasks
- [ ] Survey potential solutions from existing work. E.g., function call of GPT, Gorilla, TaskMatrix.AI. See if we can achieve it during a straightforward solution, e.g., GPT4 + function call, or solutions from existing work.
```

### PR DESCRIPTION
**Pull request by thinkall**
## Why are these changes needed?

RetrieveChat is a convesational framework for retrieve augmented code generation and question answering. Essentially,`RetrieveAssistantAgent` and  `RetrieveUserProxyAgent` implements a different auto reply mechanism corresponding to the RetrieveChat prompts.

## Related issue number

Closes #1107 

## Checks

- [ ] I've included any doc changes needed for https://microsoft.github.io/FLAML/. See https://microsoft.github.io/FLAML/docs/Contribute#documentation to build and test documentation locally.
- [ ] I've added tests (if relevant) corresponding to the changes introduced in this PR.
- [ ] I've made sure all auto checks have passed.

### CODE REVIEW

**Review by sonichi — COMMENTED** [review]
I reviewed all except the notebook. The change to the notebook can be made in the next round.
Overall, the application looks interesting and general. The implementation is creative. It's worth considering writing a blog post about it.

File: flaml/autogen/agentchat/contrib/retrieve_user_proxy_agent.py [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+    def get_max_tokens(model="gpt-3.5"):
+        if "gpt-4-32k" in model:
+            return 32000
+        elif "gpt-4" in model:
+            return 8000
+        else:
+            return 4000
+
+    def _reset(self):
+        self._oai_conversations.clear()
```
[the remark this answers is not in the document]
Comment (thinkall) [reply flaml/autogen/agentchat/contrib/retrieve_user_proxy_agent.py:110]: This is needed as we want to clean only the messages in the conversation, but not _consecutive_auto_reply_counter

File: setup.py [hunk 2]
```diff
@@ -140,6 +140,15 @@
...
         "openai": ["openai==0.27.8", "diskcache"],
         "autogen": ["openai==0.27.8", "diskcache", "docker"],
+        "retrievechat": [
+            "openai==0.27.8",
+            "diskcache",
+            "docker",
+            "chromadb",
+            "pydantic==1.10.9",
+            "tiktoken",
+            "sentence_transformers",
```
[the remark this answers is not in the document]
Comment (thinkall) [reply setup.py:150]: Yes. For embeddings.

File: flaml/autogen/agentchat/contrib/retrieve_user_proxy_agent.py [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+
+    def receive(self, message: Union[Dict, str], sender: "Agent"):
```
[the remark this answers is not in the document]
Comment (thinkall) [reply flaml/autogen/agentchat/contrib/retrieve_user_proxy_agent.py]: Overriding generate_reply() doesn't work as when human_input_mode == "ALWAYS" it'll never trigger generate_reply in receive().

**Review by sonichi — COMMENTED** [review]
Could you add a test?
Could you review #1165 ? It provides some utility for this PR.

File: flaml/autogen/agentchat/contrib/retrieve_user_proxy_agent.py [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+        return doc_contents
+
+    def _generate_message(self, doc_contents):
+        if self.customized_prompt:
+            message = self.customized_prompt + "\nUser's question is: " + self.problem + "\nContext is: " + doc_contents
+        else:
+            message = PROMPT.format(input_question=self.problem, input_context=doc_contents)
+        return message
+
+    def receive(self, message: Union[Dict, str], sender: Agent):
```
[the remark this answers is not in the document]
Comment (thinkall) [reply flaml/autogen/agentchat/contrib/retrieve_user_proxy_agent.py:150]: Without #1165, overriding generate_reply() doesn't work as when human_input_mode == "ALWAYS" it'll never trigger generate_reply in receive(). We can do it now as `_check_termination_and_human_reply` is moved to generate_reply().

File: flaml/autogen/agentchat/contrib/retrieve_assistant_agent.py [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+            str or dict or None: reply. None if no reply is generated.
+        """
+        assert messages is not None or sender is not None, "Either messages or sender must be provided."
+        if messages is None:
+            messages = self._oai_messages[sender.name]
+        message = self._message_to_dict(messages[-1])
+        if "exitcode: 0 (execution succeeded)" in message.get("content", ""):
+            return "TERMINATE"
+        elif "UPDATE CONTEXT" in message.get("content", "")[-20::].upper():
+            self._reset()
```
[the remark this answers is not in the document]
Comment (thinkall) [reply flaml/autogen/agentchat/contrib/retrieve_assistant_agent.py:85]: Without #1165, this line should be in both send() and generate_reply(). Now, it just need to be in send().

File: flaml/autogen/agentchat/contrib/retrieve_user_proxy_agent.py [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+                    [autogen.Completion.create](../oai/Completion#create).
+            sender: sender of an Agent instance.
+
+        Raises:
+            ValueError: if the message can't be converted into a valid ChatCompletion message.
+        """
+        message = self._message_to_dict(message)
+        if "UPDATE CONTEXT" in message.get("content", "")[-20::].upper():
+            print("Updating context and resetting conversation.")
+            self._reset()
```
[the remark this answers is not in the document]
Comment (thinkall) [reply flaml/autogen/agentchat/contrib/retrieve_user_proxy_agent.py:173]: Indeed. Thanks.

File: flaml/autogen/agentchat/contrib/retrieve_assistant_agent.py [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+
+        Returns:
+            str or dict or None: reply. None if no reply is generated.
+        """
+        assert messages is not None or sender is not None, "Either messages or sender must be provided."
+        if messages is None:
+            messages = self._oai_messages[sender.name]
+        message = self._message_to_dict(messages[-1])
+        if "exitcode: 0 (execution succeeded)" in message.get("content", ""):
+            return "TERMINATE"
```
[the remark this answers is not in the document]
Comment (thinkall) [reply flaml/autogen/agentchat/contrib/retrieve_assistant_agent.py:83]: Although sometimes even when the code execution succeeds, the task is not solved, but it's hard to tell. If the human_input_mode of RetrieveUserProxyAgent is "TERMINATE" or "ALWAYS", user can still continue the conversation.

File: flaml/autogen/agentchat/contrib/retrieve_user_proxy_agent.py [hunk 8]
```diff
@@ -0,0 +1,258 @@
...
+            # if result.error_before_exec is not None:
+            #     log += f"\n{result.error_before_exec}"
+            #     exitcode = 1
+            # if result.error_in_exec is not None:
+            #     log += f"\n{result.error_in_exec}"
+            #     exitcode = 1
+            # return exitcode, log, None
+
+            result = self._ipython.run_cell(code)
+            log = str(result.result)
```
[the remark this answers is not in the document]
Comment (thinkall) [reply flaml/autogen/agentchat/contrib/retrieve_user_proxy_agent.py:250]: I'm not sure whether we should capture all of the stdout and stderr outputs and include them in the message. In the case of FLAML, the training logs can often be too long to fit into the LLM tokens. The result.result can only capture concise error information, which may not be sufficient for debugging problems. Therefore, I believe that for complex problems, it's essential to include human input as well.

**Review by sonichi — COMMENTED** [review]
Most comments are for changes required in future PR. A few comments need to be addressed in this PR. If it's unclear which are required for this PR, please let me know.

File: notebook/autogen_agentchat_RetrieveChat.ipynb
Comment (sonichi) [inline notebook/autogen_agentchat_RetrieveChat.ipynb:1]: The initial message is printed twice. But I can't find the extra print statement in the code.
Using colored for some control message may improve the readability.

**Review by sonichi — COMMENTED** [review]
Almost ready. I have some minor comments.

File: notebook/autogen_agentchat_RetrieveChat.ipynb
Comment (thinkall) [reply notebook/autogen_agentchat_RetrieveChat.ipynb:1]: Very weired for the print twice issue. It only appears in notebook, but not in terminal mode. It's fixed by applying your suggestion for `_generate_retrieve_user_reply`.

**Comment by sonichi** [comment]
For future PRs that requires OpenAI test, please use a branch in microsoft/FLAML

**Comment by kevin666aa** [comment]
Since the notebook is very long, can you add a table of contents for the examples at the beginning? That is what the example demonstrate and unique features of retrieve chat. For example, Example 5: we demonstrate the unique feature "UPDATE CONTEXT"...

**Comment by thinkall** [comment]
> Since the notebook is very long, can you add a table of contents for the examples at the beginning? That is to show what the example demonstrates and unique features of retrieve chat, and the user can go to the corresponding examples. For example, Example 5: we demonstrate the unique feature "UPDATE CONTEXT"...

Thank you @kevin666aa for the suggestion. I've added the ToC, it works in VSCode and local jupyter server, but doesn't work as expected in GitHub preview. Looks like github has limited [preview support](https://docs.github.com/en/repositories/working-with-files/using-files/working-with-non-code-files#working-with-jupyter-notebook-files-on-github) on the links.

**Comment by kevin666aa** [comment]
We can just put the contents without the links. I think people can easily go to the example by searching the keywords.

**Review by kevin666aa — APPROVED** [review]

**Pull request merged**